File: frontend/src/app/services/mecanico.service/mecanico.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environment';

export interface EspecialidadDto {
  idEspecialidad: number;
  nombre: string;
}

export interface MecanicoDto {
  idMecanico: number;
  nombre: string;
  contacto: string;
  certificacion: string | null;
  edad: number;
  experiencia: number;
  ciudad: string;
  foto: string | null;
  disponible: boolean;
  rating: number | null;
  cantidadValoraciones: number;
  especialidades: EspecialidadDto[];
}

@Injectable({ providedIn: 'root' })
export class MecanicoService {
  private readonly api = environment.apiUrl;

  constructor(private http: HttpClient) {}

  getMecanicos(filters?: {
    especialidadId?: number;
    ciudad?: string;
    disponible?: boolean;
  }): Observable<MecanicoDto[]> {
    let params = new HttpParams();
    if (filters?.especialidadId != null) {
      params = params.set('especialidadId', filters.especialidadId);
    }
    if (filters?.ciudad) {
      params = params.set('ciudad', filters.ciudad);
    }
    if (filters?.disponible != null) {
      params = params.set('disponible', filters.disponible);
    }
    return this.http.get<MecanicoDto[]>(`${this.api}/mecanicos`, { params });
  }

  getMecanicoById(id: number): Observable<MecanicoDto> {
    return this.http.get<MecanicoDto>(`${this.api}/mecanicos/${id}`);
  }
}

File: backend/pist-api/src/main/java/com/pits/pist_api/service/MecanicoService.java
package com.pits.pist_api.service;

import com.pits.pist_api.dto.EspecialidadDto;
import com.pits.pist_api.dto.MecanicoDto;
import com.pits.pist_api.entity.Mecanico;
import com.pits.pist_api.repository.MecanicoRepository;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class MecanicoService {

    private final MecanicoRepository repo;

    public MecanicoService(MecanicoRepository repo) {
        this.repo = repo;
    }

    public List<MecanicoDto> listar(Long especialidadId, String ciudad, Boolean disponible) {
        boolean sinFiltros = especialidadId == null
                && (ciudad == null || ciudad.isBlank())
                && disponible == null;
        if (sinFiltros) {
            return repo.findAll().stream().map(this::toDto).toList();
        }
        return repo.findWithFilters(especialidadId, ciudad, disponible)
                .stream().map(this::toDto).toList();
    }

    public MecanicoDto obtenerPorId(Long id) {
        Mecanico m = repo.findById(id)
                .orElseThrow(() -> new RuntimeException("Mecánico no encontrado"));
        return toDto(m);
    }

    private MecanicoDto toDto(Mecanico m) {
        List<EspecialidadDto> especialidades = m.getEspecialidades().stream()
                .map(e -> EspecialidadDto.builder()
                        .idEspecialidad(e.getIdEspecialidad())
                        .nombre(e.getNombre())
                        .build())
                .toList();

        return MecanicoDto.builder()
                .idMecanico(m.getIdMecanico())
                .nombre(m.getNombre())
                .contacto(m.getContacto())
                .certificacion(m.getCertificacion())
                .edad(m.getEdad())
                .experiencia(m.getExperiencia())
                .ciudad(m.getCiudad())
                .foto(m.getFoto())
                .disponible(m.getDisponible())
                .rating(m.getRating())
                .cantidadValoraciones(m.getCantidadValoraciones())
                .especialidades(especialidades)
                .build();
    }
}
